File: ui/action-button.js
import { Controller } from "../web/stimulus.js";
import { notify } from "./note-port.js";

export default class extends Controller {
	static values = { url: String, params: Object };

	async call() {
		this.element.disabled = true;
		const body = new URLSearchParams(this.paramsValue);
		try {
			const resp = await fetch(this.urlValue, {
				method: "POST",
				body,
				headers: { Accept: "text/vnd.turbo-stream.html" },
			});
			if (resp.redirected) {
				Turbo.visit(resp.url);
				return;
			}
			if (!resp.ok) {
				notify("Something went wrong");
				this.element.disabled = false;
				return;
			}
			const ct = resp.headers.get("Content-Type") || "";
			if (ct.includes("turbo-stream")) {
				Turbo.renderStreamMessage(await resp.text());
			}
		} catch {
			notify("Could not reach the server");
		}
		this.element.disabled = false;
	}
}

File: ui/action_button.go
package ui

import (
	"encoding/json/v2"

	"ily.dev/act3/html"
	"ily.dev/act3/html/attr"
	"ily.dev/act3/ui/stimulus"
)

// ActionButton renders a button that POSTs to action
// with form-encoded params on click.
// The controller handles redirects (via Turbo.visit)
// and turbo-stream responses (via Turbo.renderStreamMessage).
func ActionButton(action string, params map[string]string, attrs ...attr.Node) html.Element {
	return func(children ...html.Node) html.Node {
		as := []attr.Node{
			stimulus.Controller("action-button"),
			stimulus.Value("action-button", "url")(action),
			stimulus.Action("click->action-button#call"),
		}
		if len(params) > 0 {
			paramsJSON, _ := json.Marshal(params)
			as = append(as, stimulus.Value("action-button", "params")(string(paramsJSON)))
		}
		as = append(as, group(attrs...))
		return Button(as...)(children...)
	}
}
